control/controller.go

```diff
@@ -632,8 +642,14 @@ TRANSITION:
 	}
 	q.currentSpan, q.currentCtx = nil, nil
 
-	// If we are transitioning to a finished state from a non-finished state, finish the parent span.
 	if isFinishedState(newState) {
+		// Invoke the cancel function to ensure that we have signaled that the query should be done.
+		// The user is supposed to read the entirety of the tables returned before we end up in a finished
+		// state, but user error may have caused this not to happen so there's no harm to canceling multiple
+		// times.
+		q.cancel()
```

Comment:
Was this a previous bug in the controller?

Reply:
It might have been? I don't think it could have deadlocked previously. Now, the `Done()` method could get blocked on the `es.do()` invocation since the metadata channel needs to be closed.